README.md
# jQuery Numeric Input [![Build Status](https://secure.travis-ci.org/manuelvanrijn/jquery-numeric_input.png?branch=master)][travis]

[travis]: http://travis-ci.org/manuelvanrijn/jquery-numeric_input

This plugins helps blocking input fields only to except numeric value's. Also it provides the function to transform the dot or comma input to a char you provide.

## Getting Started
Download the [production version][min] or the [development version][max].

[min]: https://raw.github.com/manuelvanrijn/jquery-numeric_input/master/dist/jquery.numeric_input.min.js
[max]: https://raw.github.com/manuelvanrijn/jquery-numeric_input/master/dist/jquery.numeric_input.js

In your web page:

```html
<script src="libs/jquery/jquery.js"></script>
<script src="dist/jquery.numeric_input.min.js"></script>
<script>
$(function() {
    $("input.numeric_input").numeric_input();
});
</script>
```

## Default options

Here's a list with the available default options

```javascript
{
    decimal: ',',           // char to append when dot or comma is pressed
    leadingZeroCheck: true, // append a 0 when the first input is a comma or dot
    initialParse: true,     // parse the value on start to match the decimal option
    parseOnBlur: true,      // parse value when user lose focus on field
    allowNegative: false    // allow value to be negative (prepends a minus char)
}
```

You can override the defaults globally. For instance, you can override the global default value for `decimal` by defining:

```javascript
$.fn.numeric_input.defaults.decimal = '.';
```

## Examples

### Change the decimal char

```javascript
// set's the decimal to a '.' instead of a ','
$("input.numeric_input").numeric_input({
  decimal: '.'
});
```

### Don't append a leading zero

```javascript
// set's the decimal to a '.' instead of a ','
$("input.numeric_input").numeric_input({
  leadingZeroCheck: false
});
```

### Don't parse the value on start

```javascript
$("input.numeric_input").numeric_input({
  initialParse: false
});
```

### Don't parse the value on blur

```javascript
$("input.numeric_input").numeric_input({
  parseOnBlur: false
});
```

### Allow negative values

```javascript
$("input.numeric_input").numeric_input({
  allowNegative: true
});
```

## Changelog

A detailed overview of can be found in the [CHANGELOG](https://github.com/manuelvanrijn/jquery-numeric_input/blob/master/CHANGELOG.md).

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint and test your code using [grunt](https://github.com/cowboy/grunt).

_Also, please don't edit files in the "dist" subdirectory as they are generated via grunt. You'll find source code in the "src" subdirectory!_

## License
Copyright (c) 2013 Manuel van Rijn
Licensed under the MIT, GPL licenses.

[![githalytics.com alpha](https://cruel-carlota.pagodabox.com/74b24b73a364e01a82ac840cacd77fb4 "githalytics.com")](http://githalytics.com/manuelvanrijn/jquery-numeric_input)


CHANGELOG.md
### TBA

### v0.1.3

- Issue #10 Allow negative values with the option `allowNegative: true`
- Issue #9  Fixes issue with trim for IE8
- Issue #6  Fixes problem with special key's not being triggered

### v0.1.2

- Issue #3 Added onBlur support so autocompleted value's will also be parsed

### v0.1.1

- Issue #1 Added support for initially parsing the value of the input

### v0.1.0

- Initial version


test/jquery.numeric_input_test.js
/*global QUnit:false, module:false, test:false, asyncTest:false, expect:false*/
/*global start:false, stop:false ok:false, equal:false, notEqual:false, deepEqual:false*/
/*global notDeepEqual:false, strictEqual:false, notStrictEqual:false, raises:false*/
(function($) {

  /*
    ======== A Handy Little QUnit Reference ========
    http://docs.jquery.com/QUnit

    Test methods:
      expect(numAssertions)
      stop(increment)
      start(decrement)
    Test assertions:
      ok(value, [message])
      equal(actual, expected, [message])
      notEqual(actual, expected, [message])
      deepEqual(actual, expected, [message])
      notDeepEqual(actual, expected, [message])
      strictEqual(actual, expected, [message])
      notStrictEqual(actual, expected, [message])
      raises(block, [expected], [message])
  */

  module('jQuery#numeric_input', {
    setup: function() {
      $('#qunit-fixture').append('<input type="text" id="numeric" />');
      this.target = $('#qunit-fixture #numeric');
    },
    teardown: function() {
      $('#qunit-fixture').empty();
      this.target = undefined;
    }
  });

  test('default options', function() {
    ok($.fn.numeric_input.defaults, 'default options defined');

    strictEqual($.fn.numeric_input.defaults.decimal, ',', 'default value for decimal check');
    strictEqual($.fn.numeric_input.defaults.leadingZeroCheck, true, 'default value for leadingZeroCheck check');
    strictEqual($.fn.numeric_input.defaults.initialParse, true, 'default value for initialParse check');
    strictEqual($.fn.numeric_input.defaults.parseOnBlur, true, 'default value for parseOnBlur check');

  });

  test('can change default options globally', function() {
    $.fn.numeric_input.defaults.decimal = '!!!';
    equal($.fn.numeric_input.defaults.decimal, '!!!', 'can change the defaults globally');
    $.fn.numeric_input.defaults.decimal = ',';
  });

  test('is chainable', 1, function() {
    $('#qunit-fixture').append('<input type="text" id="numeric_2" />');
    var targets = $('#qunit-fixture').children();
    strictEqual(targets.numeric_input(), targets, 'should be chaninable');
  });

  test('set instance', function() {
    this.target.numeric_input();
    ok(this.target.data('numeric_input'), 'should set the data property with the instance');
  });

  module('NumericInput', {
    setup: function() {
      $('#qunit-fixture')
        .append('<input type="text" id="numeric" />')
        .append('<input type="text" id="numeric_2" />');
      this.target = $('#qunit-fixture #numeric');
      this.numeric_input = $('#qunit-fixture #numeric_2').numeric_input().data('numeric_input');
    },
    teardown: function() {
      $('#qunit-fixture').empty();
    }
  });

  test('block all input except numeric', function() {
    equal(this.numeric_input.preventDefaultForKeyCode( 44 ), true, 'block comma');
    equal(this.numeric_input.preventDefaultForKeyCode( 188 ), true, 'block comma');
    equal(this.numeric_input.preventDefaultForKeyCode( 45 ), true, 'block minus');
    equal(this.numeric_input.preventDefaultForKeyCode( 46 ), true, 'block dot');
    equal(this.numeric_input.preventDefaultForKeyCode( 190 ), true, 'block dot');
    equal(this.numeric_input.preventDefaultForKeyCode( 47 ), true, 'block slash');

    equal(this.numeric_input.preventDefaultForKeyCode( 48 ), false, 'do not block 0');
    equal(this.numeric_input.preventDefaultForKeyCode( 49 ), false, 'do not block 1');
    equal(this.numeric_input.preventDefaultForKeyCode( 50 ), false, 'do not block 2');
    equal(this.numeric_input.preventDefaultForKeyCode( 51 ), false, 'do not block 3');
    equal(this.numeric_input.preventDefaultForKeyCode( 52 ), false, 'do not block 4');
    equal(this.numeric_input.preventDefaultForKeyCode( 53 ), false, 'do not block 5');
    equal(this.numeric_input.preventDefaultForKeyCode( 54 ), false, 'do not block 6');
    equal(this.numeric_input.preventDefaultForKeyCode( 55 ), false, 'do not block 7');
    equal(this.numeric_input.preventDefaultForKeyCode( 56 ), false, 'do not block 8');
    equal(this.numeric_input.preventDefaultForKeyCode( 57 ), false, 'do not block 9');
  });

  test('do not block special keys', function() {
    equal(this.numeric_input.preventDefaultForKeyCode( 0 ), false, 'do not block browser specific key');
    equal(this.numeric_input.preventDefaultForKeyCode( 8 ), false, 'do not block backspace');
    equal(this.numeric_input.preventDefaultForKeyCode( 9 ), false, 'do not block tab');
    equal(this.numeric_input.preventDefaultForKeyCode( 35 ), false, 'do not block end');
    equal(this.numeric_input.preventDefaultForKeyCode( 36 ), false, 'do not block home');
    equal(this.numeric_input.preventDefaultForKeyCode( 37 ), false, 'do not block left arrow');
    equal(this.numeric_input.preventDefaultForKeyCode( 39 ), false, 'do not block rigth arrow');
    equal(this.numeric_input.preventDefaultForKeyCode( 144 ), false, 'do not block num lock');
  });

  test('append leading zero', function() {
    equal(this.numeric_input.getNewValueForKeyCode(44, ''), '0,', 'should prepand a leading zero by default for comma');
    equal(this.numeric_input.getNewValueForKeyCode(46, ''), '0,', 'should prepand a leading zero by default for dot');

    this.target.numeric_input({
      leadingZeroCheck: false
    });

    equal(this.target.data('numeric_input').getNewValueForKeyCode(44, ''), ',', 'should NOT prepand a leading zero when specified for comma');
    equal(this.target.data('numeric_input').getNewValueForKeyCode(46, ''), ',', 'should NOT prepand a leading zero when specified for dot');
  });

  test('different decimal char', function() {
    equal(this.numeric_input.getNewValueForKeyCode(44, '987'), '987,', 'should append a comma by default for a comma');
    equal(this.numeric_input.getNewValueForKeyCode(46, '987'), '987,', 'should append a comma by default for a dot');

    this.target.numeric_input({
      decimal: '#'
    });

    equal(this.target.data('numeric_input').getNewValueForKeyCode(44, '987'), '987#', 'should append a # when specified for a comma');
    equal(this.target.data('numeric_input').getNewValueForKeyCode(46, '987'), '987#', 'should append a # when specified for a dot');
  });

  test('allow only one instance of decimal in the value', function() {
    equal(this.numeric_input.getNewValueForKeyCode(44, '12,34'), false, 'should not append a second comma when comma is pressed');
    equal(this.numeric_input.getNewValueForKeyCode(46, '12,34'), false, 'should not append a second comma when dot is pressed');

    this.target.numeric_input({
      decimal: '#'
    });

    equal(this.target.data('numeric_input').getNewValueForKeyCode(44, '12#34'), false, 'should not append a second # when comma is pressed');
    equal(this.target.data('numeric_input').getNewValueForKeyCode(46, '12#34'), false, 'should not append a second # when dot is pressed');
  });

  test('should parse the value on blur by default', function() {
    this.target.val('1.0').numeric_input({
      initialParse: false
    });
    this.target.trigger('blur');
    equal(this.target.val(), '1,0', 'should have parsed the value on blur');
  });

  test('should not parse the value on blur parseOnBlur is false', function() {
    this.target.val('1.0').numeric_input({
      initialParse: false,
      parseOnBlur: false
    });
    this.target.trigger('blur');
    equal(this.target.val(), '1.0', 'should not have parsed the value on blur');
  });

  module('Initial value', {
    setup: function() {
      $('#qunit-fixture').append('<input type="text" id="numeric" />');
      this.target = $('#qunit-fixture #numeric');
    },
    teardown: function() {
      $('#qunit-fixture').empty();
      this.target = undefined;
    }
  });

  test('should parse the initial value with default options', function() {
    this.target.val('1.0').numeric_input();
    equal(this.target.val(), '1,0', 'should parsed the value to match the ',' decimal');
  });

  test('should parse the initial value with other decimal option', function() {
    this.target.val('1,0').numeric_input({
      decimal: '#'
    });
    equal(this.target.val(), '1#0', 'should parsed the value to match the # decimal');
  });

  test('should parse the initial value if it starts with a .', function() {
    this.target.val('.25').numeric_input();
    equal(this.target.val(), '0,25', 'should parsed the value and append the 0');
  });

  test('should parse the negative value', function() {
    this.target.val('-2.0').numeric_input({
      allowNegative: true
    });
    equal(this.target.val(), '-2,0', 'should parsed the value');
  });

  test('should parse the negative value without leading zero', function() {
    this.target.val('-.3').numeric_input({
      allowNegative: true
    });
    equal(this.target.val(), '-0,3', 'should parsed the value');
  });

  module('Skip initial parse', {
    setup: function() {
      $('#qunit-fixture').append('<input type="text" id="numeric" />');
      this.target = $('#qunit-fixture #numeric');
    },
    teardown: function() {
      $('#qunit-fixture').empty();
      this.target = undefined;
    }
  });

  test('should not parse the initial value of option initialParse is false for 1.0', function() {
    this.target.val('1.0').numeric_input({
      initialParse: false
    });
    equal(this.target.val(), '1.0', 'should not parse the value to 1,0');
  });

  test('should not parse the initial value of option initialParse is false for .25', function() {
    this.target.val('.25').numeric_input({
      initialParse: false
    });
    equal(this.target.val(), '.25', 'should not parse the value to  0.25');
  });

  module('Cleanup string', {
    setup: function() {
      $('#qunit-fixture').append('<input type="text" id="numeric" />');
      this.target = $('#qunit-fixture #numeric');
      this.numeric_input = this.target.numeric_input({
        allowNegative: true
      }).data('numeric_input');
    },
    teardown: function() {
      $('#qunit-fixture').empty();
      this.target = undefined;
      this.numeric_input = undefined;
    }
  });

  test('should strip unsupported chars', function() {
    equal(this.numeric_input.parseValue("abcdefghijklmnopqrstuvwxyz"), '');
    equal(this.numeric_input.parseValue("ABCDEFGHIJKLMNOPQRSTUVWXYZ"), '');
    equal(this.numeric_input.parseValue("a1b2c3D4E5F6Z"), '123456');
    equal(this.numeric_input.parseValue("123456789.987654321,123456789"), '123456789,987654321,123456789');
    equal(this.numeric_input.parseValue("-123,456"), '-123,456');
  });

  module('Allow negative values', {
    setup: function() {
      $('#qunit-fixture')
        .append('<input type="text" id="numeric" />')
        .append('<input type="text" id="numeric_2" />');
      this.target = $('#qunit-fixture #numeric');
      this.second_target = $('#qunit-fixture #numeric_2');
      this.numeric_input = this.target.numeric_input({
        allowNegative: true
      }).data('numeric_input');
    },
    teardown: function() {
      $('#qunit-fixture').empty();
      this.target = undefined;
      this.second_target = undefined;
      this.numeric_input = undefined;
    }
  });

  test('should prepend the minus char if allowed', function() {
    equal(this.numeric_input.getNewValueForKeyCode(45, '2,0'), '-2,0', 'should prepand the minus');
    equal(this.numeric_input.getNewValueForKeyCode(45, '-2,0'), false, 'should not prepand the minus if already prepended');
    equal(this.numeric_input.getNewValueForKeyCode(45, ''), '-', 'should prepend the minus');
  });

  test('should not allow a negative value if allowNegative is false', function() {
    this.second_target.val('-1.0').numeric_input({
      allowNegative: false
    });
    equal(this.second_target.val(), '1,0', 'should parse the value to 1,0');
  });

  test('should not allow a negative value if allowNegative is false', function() {
    this.second_target.val('-1.0').numeric_input({
      allowNegative: true
    });
    equal(this.second_target.val(), '-1,0', 'should parse the value to 1,0');
  });

}(jQuery));


grunt.js
/*global module:false*/
module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    pkg: '<json:package.json>',
    meta: {
      banner: '/*\n' +
        ' * <%= pkg.title || pkg.name %> - v<%= pkg.version %> - ' +
        '<%= grunt.template.today("yyyy-mm-dd") %>\n' +
        '<%= pkg.homepage ? " * " + pkg.homepage + "\n" : "" %>' +
        ' * Copyright (c) <%= grunt.template.today("yyyy") %> <%= pkg.author.name %>\n' +
        ' * Licensed <%= _.pluck(pkg.licenses, "type").join(", ") %>\n' +
        ' */'
    },
    concat: {
      dist: {
        src: ['<banner:meta.banner>', '<file_strip_banner:src/<%= pkg.name %>.js>'],
        dest: 'dist/<%= pkg.name %>.js'
      }
    },
    min: {
      dist: {
        src: ['<banner:meta.banner>', '<config:concat.dist.dest>'],
        dest: 'dist/<%= pkg.name %>.min.js'
      }
    },
    qunit: {
      files: ['test/**/*.html']
    },
    lint: {
      files: ['grunt.js', 'src/**/*.js', 'test/**/*.js']
    },
    watch: {
      files: '<config:lint.files>',
      tasks: 'lint qunit'
    },
    jshint: {
      options: {
        curly: true,
        eqeqeq: true,
        immed: true,
        latedef: true,
        newcap: true,
        noarg: true,
        sub: true,
        undef: true,
        boss: true,
        eqnull: true,
        browser: true
      },
      globals: {
        jQuery: true
      }
    },
    uglify: {}
  });

  // Default task.
  grunt.registerTask('default', 'lint qunit concat min');

  // Travis CI task.
  grunt.registerTask('travis', 'lint qunit');
};


src/jquery.numeric_input.js
;(function( $, window, document, undefined ) {

  var NumericInput = function( elem, options ) {
    this.elem = elem;
    this.$elem = $(elem);
    this.options = options;
  };

  NumericInput.prototype = {

    init: function() {
      // set instance variable
      var _instance = this;

      _instance.options = $.extend({}, $.fn.numeric_input.defaults, _instance.options);

      // bind the keypress event
      _instance.$elem.keypress(function( e ) {
        if( _instance.preventDefaultForKeyCode(e.which) === true) {
          e.preventDefault();
        }
        var newValue = _instance.getNewValueForKeyCode( e.which, _instance.$elem.val() );
        if( newValue !== false) {
          _instance.$elem.val( newValue );
        }
      });

      if( _instance.options.parseOnBlur === true ) {
        // bind the blur event to (re)parse value
        _instance.$elem.blur(function( e ) {
          var parsedValue = _instance.parseValue( _instance.$elem.val() );
          _instance.$elem.val( parsedValue );
        });
      }

      // initial parse values
      if( _instance.options.initialParse === true ) {
        var parsedValue = _instance.parseValue( _instance.$elem.val() );
        _instance.$elem.val( parsedValue );
      }

      return _instance;
    },

    preventDefaultForKeyCode: function( keyCode ) {
      // numeric
      if( keyCode >= 48 && keyCode <= 57) {
        return false;
      }
      // special key's
      switch ( keyCode ) {
        case 0 :     // browser specific special key
        case 8 :     // backspace
        case 9 :     // tab
        case 35 :    // end
        case 36 :    // home
        case 37 :    // left
        case 39 :    // right
        case 144 :   // num lock
          return false;
        default:
          return true;
      }
    },

    getNewValueForKeyCode: function( keyCode, currentValue ) {
      // if a comma or a dot is pressed...
      if( keyCode === 44 || keyCode === 46 || keyCode === 188 || keyCode === 190 ) {
        // and we do not have a options.decimal present...
        if( currentValue.indexOf( this.options.decimal ) === -1 ) {
          // append leading zero if currentValue is empty and leadingZeroCheck is active
          if( $.trim(currentValue) === '' && this.options.leadingZeroCheck ) {
            currentValue = '0';
          }
          // append the options.decimal instead of the dot or comma
          return currentValue + this.options.decimal;
        }
      }
      return false;
    },

    parseValue: function( value ) {
      var result = value;
      if ( result.indexOf('.') !== -1 || result.indexOf(',') !== -1 ) {
        result = result.replace( '.', this.options.decimal );
        result = result.replace( ',', this.options.decimal );
      }
      if ( result.indexOf( this.options.decimal ) === 0 ) {
        result = '0' + result;
      }
      return result;
    }
  };

  $.fn.numeric_input = function ( options ) {
    return this.each(function () {
      if ( !$.data(this, 'numeric_input') ) {
        $.data( this, 'numeric_input', new NumericInput(this, options).init() );
      }
    });
  };

  $.fn.numeric_input.defaults = {
    decimal: ',',
    leadingZeroCheck: true,
    initialParse: true,
    parseOnBlur: true
  };

}( jQuery, window, document ));
